### PR TITLE
www/nginx: Log display improvements

### PR DESCRIPTION
**Pull request by 8191**
* Shift complete log display logic from client to server to allow efficient display of big logfiles.
  * Implemented server-side pagination
  * Implemented server-side filtering
* Smaller UI improvements
  * Fixed header/footer in log view
  * Prev/Next and First/Last pagination buttons added and display of number of pages
  * Display number of logs currently displayed/available/filtered, etc.
* Implement possibility to display historic, already rotated logs via web UI.
  * Re-structure menu in UI
  * Pass file-number to display to backend

image

### CODE REVIEW

**Comment by 8191** [comment]
@fabianfrz Could you please have a look at this rather big PR? I think it needs some more work as the other open PRs might conflict with some files here, but the functionality itself is ready for review/commenting. Thanks!

File: www/nginx/src/opnsense/mvc/app/library/OPNsense/Nginx/LogParserBase.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+     * Reverse read logfile, count lines, save lines which match filter and
+     * count matching lines.
+     */
+    private function parse_file()
+    {
+        $handle = @fopen($this->file_name, 'r');
+        if ($handle) {
+            $cnt = 0;
+            $total = 0;
+            for ($line = '', $pos = -2; fseek($handle, $pos, SEEK_END) !== -1; $pos--) {
```
Comment (8191) [inline www/nginx/src/opnsense/mvc/app/library/OPNsense/Nginx/LogParserBase.php:66]: The PHP implementation of `SEEK_END` seems to slow the file reading dramatically (approx. 600ms vs. 70ms when reading the complete file linear).
As one nginx log file typically is not huge in size, maybe reading the complete file to memory is fine as well.

File: www/nginx/src/opnsense/mvc/app/library/OPNsense/Nginx/LogParserBase.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                }
+
+                if ($pass) {
+                    if (floor($cnt / $this->per_page) == $this->page) {
+                        $this->result[] = $parsed_line;
+                    }
+                    $cnt++;
+                }
+            }
+
```
Comment (8191) [inline www/nginx/src/opnsense/mvc/app/library/OPNsense/Nginx/LogParserBase.php:89]: add `unset($lines)` to free memory of potential big logfile

**Review by fabianfrz — CHANGES_REQUESTED** [review]
my fault, I never published the review :(

File: www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js [hunk 3]
```diff
@@ -33,29 +33,29 @@ const LogView = Backbone.View.extend({
...
-        this.listenTo(this.filter_model, "change", this.clear_and_render);
+        this.listenTo(this.collection, "sync", this.render);
+        this.listenTo(this.collection, "update", this.render);
+        this.listenTo(this.collection.filter_model, "change", this.render);
         this.type = '';
     },
 
     render: function() {
-        let tbody = this.$el.find('tbody');
+        let tbody = this.$('tbody');
```
[the remark this answers is not in the document]
Comment (8191) [reply www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js:54]: @fabianfrz Actually if I read the Backbone docs correctly, `$` does the same here:

> If jQuery is included on the page, each view has a $ function that runs queries scoped within the view's element.

_Source: https://backbonejs.org/#View-dollar_

File: www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js [hunk 4]
```diff
@@ -65,55 +65,101 @@ const LogView = Backbone.View.extend({
...
+                this.collection.forEach(
+                    (model) => this.render_one(tbody, model)
+                );
+            }
+
+            this.$('#entrycountdisplay').html(this.page_entry_count);
+            this.$('#currentpage').html(this.current_page + 1);
+            this.$('#pagecount').html(this.collection.page_count);
+            this.$('#totalcount').html(this.collection.total_entries);
+            this.$('#resultcount').html(this.collection.displayed_entries);
```
[the remark this answers is not in the document]
Comment (8191) [reply www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js:77]: If my assumption above is correct, then this should also find the HTML elements by ID _within_ the view...

File: www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/Api/LogsController.php [hunk 5]
```diff
@@ -68,87 +77,164 @@ public function tls_handshakesAction()
...
+     * @param $page int pagination page to retrieve
+     * @param $perPage int number of entries per page
+     * @param $query string filter string to apply
+     * @return |null
+     * @throws \Exception ?
+     */
+    private function get_logs($type, $uuid, $fileno, $page, $perPage, $query)
+    {
+        if (!($this->vhost_exists($uuid) || $uuid == 'global')) {
+        // TODO check $fileno
```
[the remark this answers is not in the document]
Comment (8191) [reply www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/Api/LogsController.php:173]: I've removed that comment as this case is actually handles within the configd script (it's easier there, as here in the web-frontend we do not have FS access).

File: www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/Api/LogsController.php [hunk 6]
```diff
@@ -68,87 +77,164 @@ public function tls_handshakesAction()
...
+            $this->response->setStatusCode(404, "Not Found");
+        }
+
+        $page = intval($page);
+        $perPage = intval($perPage);
+        $query = str_replace('"', '\"', urldecode($query));
+
+        return $this->sendConfigdToClient("nginx log $type $uuid $fileno $page $perPage $query");
```
[the remark this answers is not in the document]
Comment (8191) [reply www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/Api/LogsController.php:180]: Thanks! Nice catch!
I've encoded the complete query in base64 to fix that.

File: www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabianfrz) [reply www/nginx/src/opnsense/www/js/nginx/src/controller/LogView.js:55]: Ok, I thought about $el, but sure you are right.

**Review by fabianfrz — APPROVED** [review]

**Review by fichtner — APPROVED** [review]
looks ok, @fabianfrz will you merge or shall I?

**Comment by 8191** [comment]
I've updated the release notes (b80234f) - added the change to 1.24 as it seems still not released.

**Comment by fichtner** [comment]
1.24 is out so we need 1.25 for this https://github.com/opnsense/changelog/blob/9f38b6d268a06558b2eea3efe8196ab217866d71/community/21.7/21.7.6#L34 :)

**Comment by 8191** [comment]
@fichtner Oh! Thanks to point that out. Changed the release notes in 1d82b91.

**Comment by 8191** [comment]
@fichtner Oh! Thanks to point that out! I've updated the release notes and plugin version (c9c7362).

File: www/nginx/pkg-descr [hunk 7]
```diff
@@ -10,6 +10,9 @@ WWW: https://nginx.org/
 Plugin Changelog
 ================
 
+1.25
```
Comment (fichtner) [inline www/nginx/pkg-descr:13]:
```suggestion
1.25

```

**Comment by fichtner** [comment]
Merged, thanks!

**Comment by kulikov-a** [comment]
Thanks a lot to all. looks great!

**Pull request merged**